src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -420,42 +419,59 @@ private void WriteGuid(Guid value)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private void Write(ReadOnlySpan<char> value)
         {
-            var buffer = _output.Buffer;
-            int written;
-            while (!_output.Encoder.TryEncode(value, buffer, out int consumed, out written))
-                buffer = EnsureBuffer();
+            ReadOnlySpan<byte> source = value.AsBytes();
 
-            _output.Advance(written);
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void WriteJsonValue(ReadOnlySpan<char> values)
-        {
             if (UseFastUtf8)
             {
-                int len = values.Length;
-                var buffer = EnsureBuffer(len);
+                Span<byte> destination = _output.Buffer;
 
-                ref byte utf8Bytes = ref buffer.DangerousGetPinnableReference();
-                ref char chars = ref values.DangerousGetPinnableReference();
-                for (var i = 0; i < len; i++)
-                    Unsafe.Add(ref utf8Bytes, i) = (byte)Unsafe.Add(ref chars, i);
+                while (true)
+                {
+                    var status = Encoders.Utf8.ConvertFromUtf16(source, destination, out int consumed, out int written);
+                    if (status == Buffers.TransformationStatus.Done)
+                    {
+                        _output.Advance(written);
+                        return;
+                    }
+
+                    if (status == Buffers.TransformationStatus.DestinationTooSmall)
+                    {
+                        destination = EnsureBuffer();
+                        continue;
```

Comment:
The loop just wraps here. But I think we partially encode when the destination is too small. Can we somehow not completly redo the encoding? Would it be perf positive?